src/scvi/external/mrvi/_module.py

```diff
@@ -0,0 +1,589 @@
+from __future__ import annotations
+
+from typing import Any, Callable
+
+import flax.linen as nn
+import jax
+import jax.numpy as jnp
+import numpyro.distributions as dist
+
+from scvi import REGISTRY_KEYS
+from scvi.distributions import JaxNegativeBinomialMeanDisp as NegativeBinomial
+from scvi.external.mrvi._components import AttentionBlock, Dense
+from scvi.module.base import JaxBaseModuleClass, LossOutput, flax_configure
+
+DEFAULT_PX_KWARGS = {
+    "n_hidden": 32,
+    "stop_gradients": False,
+    "stop_gradients_mlp": True,
+    "dropout_rate": 0.03,
+}
+DEFAULT_QZ_ATTENTION_KWARGS = {
+    "use_map": True,
+    "stop_gradients": False,
+    "stop_gradients_mlp": True,
+    "dropout_rate": 0.03,
+}
+DEFAULT_QU_KWARGS = {}
+
+# Lower stddev leads to better initial loss values
+_normal_initializer = jax.nn.initializers.normal(stddev=0.1)
+
+
+class _DecoderZXAttention(nn.Module):
+    """Attention-based decoder.
+
+    Parameters
+    ----------
+    n_in
+        Number of input features.
+    n_out
+        Number of output features.
+    n_batch
+        Number of batches.
+    n_latent_sample
+        Number of latent samples.
+    h_activation
+        Activation function for the output layer.
+    n_channels
+        Number of channels in the attention block.
+    n_heads
+        Number of heads in the attention block.
+    dropout_rate
+        Dropout rate.
+    stop_gradients
+        Whether to stop gradients to ``z``.
+    stop_gradients_mlp
+        Whether to stop gradients to the MLP in the attention block.
+    training
+        Whether the model is in training mode.
+    n_hidden
+        Number of hidden units in the MLP.
+    n_layers
+        Number of layers in the MLP.
+    low_dim_batch
+        Whether to use low-dimensional batch embeddings.
+    activation
+        Activation function for the MLP.
+    """
+
+    n_in: int
+    n_out: int
+    n_batch: int
+    n_latent_sample: int = 16
+    h_activation: Callable[[jax.typing.ArrayLike], jax.Array] = nn.softmax
+    n_channels: int = 4
+    n_heads: int = 2
+    dropout_rate: float = 0.1
+    stop_gradients: bool = False
+    stop_gradients_mlp: bool = False
+    training: bool | None = None
+    n_hidden: int = 32
+    n_layers: int = 1
+    training: bool | None = None
+    low_dim_batch: bool = True
+    activation: Callable[[jax.typing.ArrayLike], jax.Array] = nn.gelu
+
+    @nn.compact
+    def __call__(
+        self,
+        z: jax.typing.ArrayLike,
+        batch_covariate: jax.typing.ArrayLike,
+        size_factor: jax.typing.ArrayLike,
+        continuous_covariates: jax.typing.ArrayLike | None,
```

Comment:
This argument is currently not being used in the forward pass - is the plan to add support for this or should we remove it?

Reply:
yes let's remove, i can do this